Ready for 2.* version of iOS and 3.* version of Android OneSignal SDK.

JKalash opened the pull request:
- Android 3.\* support.
- Support for OSNotification / OSNotificationOpenResult objects.
- Added onNotificationsReceived event.
- Added syncHashedEmail & setLogLevel methods.
- Set SDK Type to "react".
- Fixed null bridge issue.
- Updated README.

Commit:
Ready for 2.* version of the SDK

Commit:
Ready for 2.* version of the SDK

Commit:
- Android 3.* support.
- Support for OSNotification / OSNotificationOpenResult objects.
- Added onNotificationsReceived event.
- Added syncHashedEmail & setLogLevel methods.
- Set SDK Type to "react".
- Updated README.

avishayil commented on index.js:
```diff
@@ -4,19 +4,26 @@
 'use strict';
 
-import { NativeModules, DeviceEventEmitter, NetInfo, Platform } from 'react-native';
+import { NativeModules, NativeEventEmitter, NetInfo, Platform } from 'react-native';
```
Make Sure you're including DeviceEventEmitter, otherwise it throws an error. You're using it to add listeners, did you mean to add the listeners using NativeEventEmitter or DeviceEventEmitter?.

avishayil commented:
Another error: issues with imports from the OneSignal SDK.
![Error](http://i.imgur.com/NV4IjnN.png][http://i.imgur.com/NV4IjnN.png)

avishayil commented on android/src/main/java/com/geektime/reactnativeonesignal/NotificationReceivedHandler.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Created by Avishay on 1/31/16.
+ */
+public class NotificationOpenedHandler implements OneSignal.NotificationReceivedHandler {
```
Probably meant here: `public class NotificationReceivedHandler`?

JKalash commented:
Pushing another commit with this fix. Also need to add support for passing init settings (InAppAlerts + AutoPrompt)

Commit:
Added inFocusDisplay method for Android. Fixed NotificationReceivedHandler class naming

JKalash commented:
@avishayil fixed in last [commit](https://github.com/OneSignal/react-native-onesignal/commit/f35064735996ffb9f186d07f7b543e8d25bfa721).

avishayil commented:
- 3.1.0 Android SDK missing methods and typos:
  1. Typo on `setlogLevel` - see fix.
  2. Remove method `enableNotificationsWhenActive` calls from `RNOneSignal.java` since it was removed from the SDK.
  3. Missing method `removeNotificationReceivedHandler` on `RNOneSignal.java`.
  4. Typo fixes on `NotificationReceivedHandler.java` - `public  class NotificationReceived` should be `public class NotificationReceivedHandler`.
  5. Missing imports on JS - `DeviceEventEmitter should be imported from React Native`
  6. Typo fixes on JS - `Notification` should be `Notifications`.

avishayil commented on android/src/main/java/com/geektime/reactnativeonesignal/NotificationReceivedHandler.java:
```diff
@@ -0,0 +1,54 @@
+
+	private ReactContext mReactContext;
+
+	public NotificationReceived(ReactContext reactContext) {
```
Probably meant `NotificationReceivedHandler`?

avishayil commented:
Please resolve all issues for compilation, then we'll proceed testing functionality.

avishayil commented on android/src/main/java/com/geektime/reactnativeonesignal/RNOneSignal.java:
```diff
@@ -118,6 +117,16 @@ public void promptLocation() {
+
+    @ReactMethod
+    public void setlogLevel(OneSignal.LOGLEVEL logLevel, OneSignal.LOGLEVEL visualLogLevel) {
+        OneSignal.setlogLevel(loglevel, visualLogLevel);
```
Typo fix:

```
    public void setlogLevel(OneSignal.LOG_LEVEL logLevel, OneSignal.LOG_LEVEL visualLogLevel) {
        OneSignal.setLogLevel(logLevel, visualLogLevel);
    }
```

avishayil commented:
You can see all the fixes on this branch:
https://github.com/geektimecoil/react-native-onesignal/tree/upgrade
Still needs fixing:
- Missing method `removeNotificationReceivedHandler` on `RNOneSignal.java` - Add the method to the SDK or remove it from `RNOneSignal.java` file.

avishayil commented:
Android Error: `C++ Exception: Got unknown argument class: OSInFocusDisplayOption.`

avishayil commented:
iOS Error:

```
Aug 30 17:02:11 Avishay-Bassas-iPhone Geektime[1049] <Error>: +[RCTOneSignal didReceiveRemoteNotification:]: unrecognized selector sent to class 0x3e2514
```

Stack Trace:

```
Aug 30 17:02:11 Avishay-Bassas-iPhone Geektime[1049] <Error>: +[RCTOneSignal didReceiveRemoteNotification:]: unrecognized selector sent to class 0x3e2514
Aug 30 17:02:12 Avishay-Bassas-iPhone com.apple.xpc.launchd[1] (UIKitApplication:com.zuznow.geektime[0x21e1][1049]) <Notice>: Service exited due to signal: Abort trap: 6
Aug 30 17:02:12 Avishay-Bassas-iPhone SpringBoard[48] <Warning>: HW kbd: Failed to set (null) as keyboard focus
Aug 30 17:02:12 Avishay-Bassas-iPhone SpringBoard[48] <Warning>: UNNotificationRegistrarConnectionListener connection invalidated
Aug 30 17:02:12 Avishay-Bassas-iPhone SpringBoard[48] <Warning>: Application 'UIKitApplication:com.zuznow.geektime[0x21e1]' crashed.
Aug 30 17:02:12 Avishay-Bassas-iPhone UserEventAgent[23] <Warning>: 16504036621393: id=com.zuznow.geektime pid=1049, state=0
Aug 30 17:02:12 Avishay-Bassas-iPhone ReportCrash[1052] <Notice>: platform_thread_get_unique_id matched 1679461
Aug 30 17:02:12 Avishay-Bassas-iPhone ReportCrash[1052] <Notice>: Formulating report for corpse[1049] Geektime
Aug 30 17:02:12 Avishay-Bassas-iPhone ReportCrash[1052] <Warning>: Saved type '109(109_Geektime)' report (8 of max 25) at /var/mobile/Library/Logs/CrashReporter/Geektime-2016-08-30-170212.ips
```

JKalash commented:
We will be pushing an Android release with the `removeNotificationReceivedHandler` method. You removed the @aar which may be cause the revert to the older version of the Android SDK, causing the Android error `C++ Exception: Got unknown argument class: OSInFocusDisplayOption.` Will push the changes.

Commit:
- Added back the @aar to prevent react from adding a dependency to 2.* instead of 3.*.
- Added method in RNOneSignal.m for didReceiveRemoteNotification
- Added changes made in /geektimecoil/react-native-onesignal/tree/upgrade

JKalash commented on ios/RCTOneSignal/RCTOneSignal.m:
```diff
@@ -60,6 +60,11 @@ - (id)initWithLaunchOptions:(NSDictionary *)launchOptions appId:(NSString *)appI
 }
 
+// This isn't required, the iOS native SDK already hooks into this event.
++ (void)didReceiveRemoteNotification:(NSDictionary *)dictionary {
```
Fixes the unrecognized selector issue

avishayil replied:
Why is this happening, if this isn't required and the iOS native SDK already hooks into this event?

JKalash commented on android/src/main/java/com/geektime/reactnativeonesignal/RNOneSignal.java:
```diff
@@ -201,6 +195,8 @@ public String getName() {
         OneSignal.removeNotificationOpenedHandler();
+
+        //Added to latest Android release
         OneSignal.removeNotificationReceivedHandler();
```
Will be added to latest release in Android native SDK

avishayil replied:
When is the Android SDK plans to be updated?
I suggest waiting until we figure out all the errors here...

avishayil commented:
Hi,
Thanks for the fixes.
- Seems that the C++ Exception on android still persists even when it's downloading the latest 3.1.0 aar:

```
Download https://jcenter.bintray.com/com/onesignal/OneSignal/3.1.0/OneSignal-3.1.0.pom
Download https://jcenter.bintray.com/com/onesignal/OneSignal/3.1.0/OneSignal-3.1.0.aar
...
:react-native-onesignal:prepareComOnesignalOneSignal310Library UP-TO-DATE
```

Stack trace:

```
C++ Exception in 'nativeRequireModuleConfig': java.lang.RuntimeException: Got unknown argument class: OSInFocusDisplayOption
get
    NativeModules.js:52
<unknown>
    index.js:10
loadModuleImplementation
    require.js:122
guardedLoadModule
    require.js:65
_require
    require.js:49
<unknown>
    AppStore.js:9
loadModuleImplementation
    require.js:122
guardedLoadModule
    require.js:65
_require
    require.js:49
<unknown>
    Navigate.js:2
loadModuleImplementation
    require.js:122
guardedLoadModule
    require.js:65
_require
    require.js:49
<unknown>
    index.android.js:6
loadModuleImplementation
    require.js:122
guardedLoadModule
    require.js:58
_require
    require.js:49
global code
    require-0.js:1
```
- The latest version hadn't been updated yet with the new `OneSignal.removeNotificationReceivedHandler()` method.

Still unable to test android properly because of the C++ bug, will test the iOS after the selector bug you fixed.

avishayil commented:
iOS: events are not dispatching.
`- (void)handleRemoteNotificationReceived:(NSString *)notification {` is triggering on the native side correctly but seems that `[self.bridge.eventDispatcher sendAppEventWithName:@"remoteNotificationReceived" body:notification];` is not triggering the event on RN.

I noticed that your `supportedEvents` array is not included anywhere else in the code, could it be that you forgot to place an observer for the events? take a look on the old code that worked, you'll notice that we put observers for the events on the `setBridge` function:

``` objc
-(void)setBridge:(RCTBridge *)receivedBridge {
   BOOL firstBridge = (_bridge == nil);
   _bridge = receivedBridge;

   if (launchDict) {
       NSLog(@"launchDict:%@", launchDict);
       [_bridge.eventDispatcher sendDeviceEventWithName:@"remoteNotificationOpened" body:launchDict];
       launchDict = nil;
   }

   if (firstBridge) {
       [[NSNotificationCenter defaultCenter] addObserver:self
                                               selector:@selector(handleRemoteNotificationReceived:)
                                                   name:OSRemoteNotificationReceived
                                                   object:nil];
       [[NSNotificationCenter defaultCenter] addObserver:self
                                               selector:@selector(handleRemoteNotificationsRegistered:)
                                                   name:OSRemoteNotificationsRegistered
                                                   object:nil];
   }
}
```

You might want to take the same approach, or find another way.

avishayil commented on index.js:
```diff
@@ -25,18 +32,27 @@ var DEVICE_IDS_AVAILABLE = 'idsAvailable';
+	if ( typeof options.onNotificationReceived !== 'undefined' ) {
+		this.onNotificationReceived = options.onNotificationReceived;
+
+		if (_pendingNotifications.length > 0) {
```
Typo: you probably meant `_pendingNotificationsReceived` here instead of `_pendingNotifications`.

avishayil commented:
I think I made some progress here, maybe you can help.
Look on this:
https://github.com/geektimecoil/react-native-onesignal/blob/update-ios-sdk/ios/RCTOneSignal/RCTOneSignal.h
https://github.com/geektimecoil/react-native-onesignal/blob/update-ios-sdk/ios/RCTOneSignal/RCTOneSignal.m

I managed to dispatch the event by subclassing `RCTEventEmitter`, but i'm getting:

```
2016-08-31 19:13:08.797 Geektime[1218:1882861] *** Assertion failure in -[RCTEventEmitter sendEventWithName:body:](), /Users/avishay/geektime-app/node_modules/react-native/React/Modules/RCTEventEmitter.m:42
libc++abi.dylib: terminate_handler unexpectedly threw an exception
```

avishayil commented:
Another issue I noticed - If I have multiple notifications waiting on the notification center, pressing on one of them dismisses them all. Need to fix this.

JKalash commented:
Hi,

We still haven't released the Android update with the method implemented which is why you're facing the Android crash. As for iOS, the `supportedEvents` is a method that you **must** override when subclassing `RCTEventEmitter`. Will have a look at the changes you made and let you know.

As for the notification dismissal, this is the expected behavior. Clicking on a notification triggers the launch of an app (or brought to foreground) which in turn clears the notification center and resets the badge count.

avishayil commented:
I'm not sure that I made the correct approach there, but it seems that the `eventDispatcher` sub methods for sending events using the bridge are deprecated now in React Native and not recommended for use (You can see that here: https://github.com/facebook/react-native/issues/8714). I'm pretty sure that I'm missing something related to the bridge itself, which I removed in the example because I encountered some errors while subclassing.
Anyway, I hope you could find the working approach for this.

JKalash commented:
The android update has been [released](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/3.1.1). You should allow a couple of hours for it to appear on maven.

avishayil commented:
Great, but it still seems to crash with the same error. Did you test it?

avishayil commented:
By the way, you can base the code on this pull, which basically does the same:
https://github.com/Microsoft/react-native-code-push/pull/500

Take a look how they did it there.

fgarci03 commented:
Are there any updates on the C++ Exception?
I'd like to chime in and help if possible, but I haven't been able to understand why this happens.

From what I noticed (and somewhat confirmed in this thread) is that OneSignal v2.x is being pulled instead/together. It hits OneSignal.OSInFocusDisplayOption on RNOneSignal.java at

`
@ReactMethod
public void inFocusDisplaying(OneSignal.OSInFocusDisplayOption displayOption) {
    OneSignal.setInFocusDisplaying(displayOption);
}
` 

Do you have any more information on this, so I can try to give it a go?

fgarci03 commented:
Ok I found something...

The enums OSInFocusDisplayOption and LOG_LEVEL can't be used as class arguments. I hacked my way into it in order to make it work, but we should find a better way. Maybe a PR to OneSignal's Android SDK?

Anyway, here's the diff: https://github.com/geektimecoil/react-native-onesignal/compare/upgrade...Seedstars:upgrade

avishayil commented:
Nice catch @fgarci03, let's see if @JKalash or @jkasten2 have a better solution for this.

jkasten2 commented:
@fgarci03 @avishayil I believe you are correct as don't think react will translate for us. We will update this for inFocusDisplaying and setLogLevel.

fgarci03 commented:
Following https://github.com/geektimecoil/react-native-onesignal/issues/69#issuecomment-245216463, this still happens in this version (with inFocusDisplaying) although a but more random (I couldn't detect a pattern as with 2.x).

I initially thought it was a problem with the SDK itself, thus tried to use this branch in order to use 3.x, but I'm not sure because I can't find this problem on their github.

avishayil commented:
@JKalash @jkasten2 how are we making progress here?

JKalash commented:
@avishayil working on it. Anything from your end?

avishayil commented:
@JKalash I'll try continue working on it and trying some stuff on the next days. Very busy at work.

JKalash commented:
@avishayil just made a commit with multiple changes fixing bridging, the android issue and other minor fixes.
Turns out React removes `RCT` prefix from the native module name as such your listener's name is `OneSignal` instead of `RCTOneSignal`.

avishayil commented:
Hi,
I created a branch for working android version, according to my testing, with some fixes I made.
Also, made some fixes for iOS, haven't tested yet.
Please take a look here:
https://github.com/geektimecoil/react-native-onesignal/tree/pr-78
@JKalash @jkasten2 Can you test on your devices?
I'll be able to do so only on sunday.

avishayil commented:
I did some other work and got idsAvailable working, but other events (Registered, Opened) are still not working, the iOS part is not executing.

@JKalash @jkasten2 

Please try to look on this version:
https://github.com/geektimecoil/react-native-onesignal/tree/pr-78

avishayil commented:
I think I know where the problem is.
I was working by your new documentation and included the following line on `AppDelegate.m`:

``` objc
  [OneSignal initWithLaunchOptions:launchOptions appId:@"***"];
```

So I figured out the the inner initWithLaunchOptions does not being called this way, so I changed it to:

``` objc
  self.oneSignal = [[RNOneSignal alloc] initWithLaunchOptions:launchOptions
                                                         appId:@"***"];
```

Then I got an error.

```
[RNOneSignal initWithLaunchOptions:appId:]: unrecognized selector sent to instance 0x177926b0
```

How do I need to send initWithLaunchOptions from `AppDelegate.m`?

JKalash commented:
@avishayil Please review the comments I made on your last commit. `RCTEventEmitter` already adopts to the `RCTBridgeModule` protocol so you should not add that to the child class. Also no synthesizing any bridge as you would cause `self` to use your `NULL` bridge instead of that of its parent `RCTEventEmitter`.

JKalash commented:
Regarding your question. Not sure what you mean by inner `initWithLaunchOptions`. Also when saying `RNOneSignal`, do you mean `RCTOneSignal`? 

I noticed that the `initWithLaunchOptions` method is not calling `self = [super init];` You should call the init of `NSObject` as your class inherits from it.

Enclose the `initWithLaunchOptions` code in :

`self = [super init];`
`if(self) {``//Existing init code ....`
`}`
`return self;`

avishayil commented:
Hi,
- I changed `RCTOneSignal` to `RNOneSignal` because you said react removes the prefix form the module name.
- By saying "inner" `initWithLaunchOptions`, I meant the one inside `RCTOneSignal.m`, seems it's not getting there.

https://github.com/geektimecoil/react-native-onesignal/tree/pr-78
- I committed your suggested changes to the branch and tried to put some logs:

``` objc
- (id)initWithLaunchOptions:(NSDictionary *)launchOptions appId:(NSString *)appId settings:(NSDictionary*)settings {
    NSLog(@"Init");
    self = [super init];
    if(self) {
        NSLog(@"Has self");
        [OneSignal setMSDKType:@"react"];
        [OneSignal initWithLaunchOptions:launchOptions
                                   appId:appId
              handleNotificationReceived:^(OSNotification* notification) {
                  [self handleRemoteNotificationReceived:[notification stringify]];
              }
                handleNotificationAction:^(OSNotificationOpenedResult * result) {
                    [self handleRemoteNotificationOpened:[result stringify]];
                }
                                settings:settings];
    }
    return self;
}
```

tried to call the following in `AppDelegate.m`:

``` objc
[OneSignal initWithLaunchOptions:launchOptions appId:@"***"];
```

But the `initWithLaunchOptions` method inside `RCTOneSignal.m` in not being called (or the code is not being executed, no `NSLog` is appearing in XCode logs). While calling instead:

``` objc
  self.oneSignal = [[RNOneSignal alloc] initWithLaunchOptions:launchOptions
                                                         appId:@"***"];
```

I get the exception.
Any ideas?

JKalash commented:
Hi @avishayil ,

Looking at the pr-78 branch, I do not see any init method for RTCOneSignal class that only takes a lunchOptions dict and an appId. There is only one with a third `settings` param. 

In the AppDelegate.m you are calling OneSignal init or RNOneSignal init without settings param. It should be non of the two. It should be something like:

`[[RNOneSignal alloc] initWithLaunchOptions:launchOptions appId:@"***" settings:@{:}];`

JKalash commented:
Also not sure where the suggested changes have been made because if you look at the PR-78 branch in the [RTCOneSignal.h](https://github.com/geektimecoil/react-native-onesignal/blob/pr-78/ios/RCTOneSignal/RCTOneSignal.h) file, there is still a `<RCTBridgeModule>`.

Last, you should change the file names to `RNOneSignal` instead of `RTCOneSignal` to match the class name.

nickpunt commented:
Outdated reference to `InAppAlertNotifications` in comment, suggest replacing with `kOSSettingsKeyInFocusDisplayOption`

avishayil commented:
@JKalash @nickpunt 
I updated the repo to the changes you mentioned.
Now, i'm trying to call it with:

``` objc
  self.oneSignal = [[RNOneSignal alloc] initWithLaunchOptions:launchOptions appId:@"***" settings:@{kOSSettingsKeyInAppAlerts : @NO, kOSSettingsKeyAutoPrompt : @NO}];
```

But when it runs, i'm getting an exception:

```
2016-09-20 17:51:48.260202 Geektime[1202:634748] *** Assertion failure in -[RCTEventEmitter sendEventWithName:body:](), /Users/avishay/geektime-app/node_modules/react-native/React/Modules/RCTEventEmitter.m:42
libc++abi.dylib: terminate_handler unexpectedly threw an exception
```

Any Ideas?

Note: Problem only happening when receiving notification. on the `IdsAvailable` the `sendEventWithName` function runs just fine.

JKalash commented:
Taking a look into this issue. Should get back to you with more soon. Also updating this PR for [2.1.12](https://github.com/OneSignal/OneSignal-iOS-SDK/releases/tag/2.1.12) iOS & [3.3.0](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/3.3.0) Android

Commit:
iOS Fixes + support for iOS SDK 2.1.12

gdeglin commented:
@avishayil  -- @JKalash told me that he believes his latest commit resolves this issue. Could you confirm?

avishayil commented:
@gdeglin @JKalash I'll take a look.

JKalash commented:
@avishayil 

Did you have a chance to look at the changes from the last [commit](https://github.com/OneSignal/react-native-onesignal/commit/c9f62607266cbed95c2444ee82e5a6668dc19908)?

avishayil commented:
@JKalash Not yet, we're on holidays so it'll be probably some day next week.

weixingsun commented:
@JKalash  I've tried OneSignal:master, but it seems not works for xcode8, since the project.pbxproj is missing. How did you compile the project with xcode?

avishayil commented:
@JKalash I believe i'll have time to test this on Tuesday.

avishayil commented:
Ok, I checked it and it's still not working. @JKalash @jkasten2 can we have a skype conversation about this?

jkasten2 committed:
Fixed iOS received and opened events.

* iOS opened and received events now fire into JS.
* Also fixed Android and iOS example code in the README.

jkasten2 committed:
iOS fixes after rebase

jkasten2 committed:
Android fixes for 2.0 release

* Fixed JSON formats of opened and received callbacks.
* Fixed js onError displaying.
* Fixed on_session / registration on first open.
* Updated README

avishayil commented on android/src/main/java/com/geektime/reactnativeonesignal/RNOneSignal.java:
```diff
@@ -31,13 +32,31 @@
+    // However it seems it is also to soon to call getCurrentActivity() from the reactContext as well.
+    // This will normally succeed when onHostResume fires instead.
+    private void initOneSignal() {
+        Activity activity = mReactContext.getCurrentActivity();
```
This leads to compilation error. Please use:
`Activity activity = getCurrentActivity();`

jkasten2 replied:
@avishayil This compiles just fine, however I do so a possible race condition where `mReactContext` maybe `null` that your suggestion will fix. Updated it in the commit below.

jkasten2 committed:
Android 2.0 fixes

* Fixed open event not firing when app is cold started from a notification.
* Fixed possible race condition with mReactContext in initOneSignal().

axe312ger commented:
This PR takes way to long. We now switched to FireBase, was pretty simple to switch over. --> https://firebase.google.com/

Sorry OneSignal, you should put more manpower in your integrations 👎

rcugut commented:
@axe312ger  lol ... pondering about this too :-)

avishayil commented:
Hi @axe312ger @rcugut you can use dev branch if you like, most of the functionality are working fine, just haven't been 100% tested yet.

axe312ger commented:
Thanks for your help but that "not 100% tested yet" is the reason I can’t use it. Also we tried several times and didn’t get any OneSignal Token. Never mind, FireBase is working great -> Opt out!

> On 15 Nov 2016, at 17:24, Avishay Bassa notifications@github.com wrote:
> 
> Hi @axe312ger https://github.com/axe312ger @rcugut https://github.com/rcugut you can use dev branch if you like, most of the functionality are working fine, just haven't been 100% tested yet.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub https://github.com/geektimecoil/react-native-onesignal/pull/78#issuecomment-260689399, or mute the thread https://github.com/notifications/unsubscribe-auth/ABqBQl6NRgc1qTwhhDc76Ed3ntIiohfOks5q-dzJgaJpZM4Jv9cO.

jkasten2 commented:
@axe312ger Were you having the same issue with the main 1.X version of this SDK?

Can you make sure to remove (or disable the push part of) any other plugins with push notifications? Some other plugins override the `application:didRegisterForRemoteNotificationsWithDeviceToken:` selector in a way that interferes with OneSignal's ability to get a push token too.

Let us know the full list of plugins you have setup or if you have any other native code in your project using push notifications.

thomasMary commented:
Poke, this PR has been out there for quite a long time now, any idea when it will be production ready and merged? Thanks

deepmehtait commented:
@JKalash @avishayil: Will this PR have silent notifications handling which is mentioned in OneSignal documentation?

avishayil commented:
That's a good question. I'm waiting for the OneSignal guys to solve some annoying problem that causes notification to alert when app is open all the time with the Android on dev branch, then we'll be able to release the new version. In the meantime, if it doesn't bother you, use the dev branch.

avishayil commented:
I finally see that the problems have been solved, and #dev branch had been merged.
Thanks!
